docs/dcp2_system_design.rst

```diff
@@ -935,6 +1028,414 @@ import tool with be requested manually in #ingest-to-tdr-shared Slack channel
+staging area. The ``entity_id`` in the name of the object must be different
+from the ``entity_id`` column of any row in any table of the TDR dataset the
+staging is imported into.
+
+This type of update creates an orphaned entity so it should only be used as
+part of other updates, as described in `Add a subgraph`_, `Add a project`_ or
+`Add an entity to existing subgraphs`_.
+
+
+Add a data file
+~~~~~~~~~~~~~~~
+
+1. Write the data file to an object in the ``data`` directory of the staging
+   area.
+
+2. Write the descriptor to an object in the ``descriptors`` directory of the
+   staging area. The ``file_uuid`` property of the descriptor must be
+   different from any ``file_uuid`` property of the JSON document in the
+   ``descriptor`` column of any row in any ``…_file`` table in the TDR dataset
+   the staging area will be imported into.
+
+   The ``entity_id`` and ``version`` fields in the name of the descriptor must
+   match those in the name of the object created in the next step.
+
+3. Add the entity describing the data file as defined in `Add an entity_`.
+
+This type of update creates an orphaned data file so it should only be used as
+part of other updates, as described in `Add a subgraph`_, `Add a project`_ or
+`Add an entity to existing subgraphs`_.
+
+
+Add a subgraph
+~~~~~~~~~~~~~~
+
+1. Add any entities introduced by the subgraph as described in `Add an
+   entity`_. If the subgraph also refers to entities that were previously
+   imported, those entities should not be written to the staging area.
+
+2. Write the subgraph to an object in the ``links`` directory of the staging
+   area. The ``links_id`` field in the name of that object must be different
+   from the value of the ``links_id`` column in any row of the ``links`` table
+   of the TDR dataset the staging are is imported into.
+
+
+Add a project
+~~~~~~~~~~~~~
+
+1. Add the ``project`` entity as described in `Add an entity`_
+
+2. Add all other entities belonging to the project as described in `Add an
+   entity`_
+
+3. Add the subgraphs belonging to the project as described in `Add a
+   subgraph`_
+
+
+Add an entity to existing subgraphs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Add the entity as described in `Add an entity`_
+
+2. Update the subgraphs that reference the entity as described in `Update a
+   subgraph`_
+
+
+Add a data file to existing subgraphs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Add the data file as described in `Add a data file`_
+
+2. Update the subgraphs that reference the data file as described in `Update a
+   subgraph`_
+
+
+Remove an entity
+~~~~~~~~~~~~~~~~
+
+1. Make sure the staging area property ``is_delta`` is ``true``.
+
+2. Create an empty object in the ``metadata`` directory of the staging area.
+   The ``entity_type`` and ``entity_id`` fields in the object name must be the
+   same as those in the name of the original (the entity to be deleted), the
+   ``version`` field must be strictly higher and the name must end in
```

Comment:
Not strictly necessary but it leads to a more elegant design that's open to future extensions. It helps to distinguish between the deletion of a version of an entity and the deletion of an entity. What's described here is the deletion of the entity, and in that sense justifies a new version. If, in the future, we wanted to allow reverting a version (including reverting a deletion!) then we would specify that as a `.remove` of an existing version. By requiring a new version we also record the time of the deletion which is useful, if not important information to record.

This is how versioned buckets work in S3: the deletion is a new version and reverting to an older version works by deleting most recent version. Reverting a deletion works by deleting the delete marker, which is a distinctly new version in itself.

This more flexible and more general design comes virtually for free, because allocating a new version timestamp is a one liner in most programming languages. In fact, I recommend that you allocate the new version at the beginning of the process of populating a staging area so that all versions produced by the process are the same. If you parameterize the process (so that the version can be passed in), you can make the process idempotent. I think I've sufficiently emphasized the importance of idempotence for robustness in previous conversations.

Reply:
Addressed